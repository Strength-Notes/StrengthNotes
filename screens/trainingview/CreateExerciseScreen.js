import React from 'react';
import {
  StyleSheet,
  Text,
  View,
  TouchableOpacity,
  TextInput,
  Picker,
} from 'react-native';
import { Card } from 'react-native-elements';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { addExerciseAction } from '../../redux/actions';
import ExerciseProperties from '../../redux/ExerciseProperties';

const styles = StyleSheet.create({
  container: {
    flex: 1,
  },
  createButtonInvalid: {
    backgroundColor: 'lightgray',
  },
  createButtonText: {
    textAlign: 'center',
  },
  createButtonTextInvalid: {
    color: 'gray',
  },
});

class CreateExerciseScreen extends React.Component {
  constructor(props) {
    super(props);

    this.navigation = props.navigation;
    this.addExerciseDispatch = props.addExerciseDispatch;

    // Set defaults: needed to prevent bug when using default value
    this.state = {
      primarySelected: ExerciseProperties.WEIGHT,
      secondarySelected: ExerciseProperties.REPS,
    };
  }

  verifyInput = () => {
    const { nameInput } = this.state;

    return !!nameInput;
  };

  handleCreatePress = () => {
    const {
      nameInput,
      categoryInput,
      primarySelected,
      secondarySelected,
    } = this.state;

    this.addExerciseDispatch({
      name: nameInput,
      category: categoryInput,
      primary: primarySelected,
      secondary: secondarySelected,
    });

    this.navigation.goBack();
  };

  render() {
    const {
      nameInput,
      categoryInput,
      primarySelected,
      secondarySelected,
    } = this.state;

    const formValid = this.verifyInput();

    return (
      <View style={styles.container}>
        <Card title="Name">
          <TextInput
            onChangeText={(text) => {
              this.setState({ nameInput: text });
            }}
            value={nameInput}
            placeholder="Name"
          />
        </Card>
        <Card title="Category">
          <TextInput
            onChangeText={(text) => {
              this.setState({ categoryInput: text });
            }}
            value={categoryInput}
            placeholder="Category"
          />
        </Card>
        <Card title="Primary">
          <Picker
            selectedValue={primarySelected}
            onValueChange={(selected) => {
              this.setState({ primarySelected: selected });
            }}
            mode="dropdown"
          >
            {
              Object.values(ExerciseProperties).map((item) => (
                <Picker.Item key={item} label={item} value={item} />
              ))
            }
          </Picker>
        </Card>
        <Card title="Secondary">
          <Picker
            selectedValue={secondarySelected}
            onValueChange={(selected) => {
              this.setState({ secondarySelected: selected });
            }}
            mode="dropdown"
          >
            {
              Object.values(ExerciseProperties).map((item) => (
                <Picker.Item key={item} label={item} value={item} />
              ))
            }
          </Picker>
        </Card>
        <TouchableOpacity
          onPress={this.handleCreatePress}
          disabled={!formValid}
        >
          <Card
            containerStyle={!formValid ? styles.createButtonInvalid : null}
          >
            <Text
              style={[
                styles.createButtonText,
                !formValid ? styles.createButtonTextInvalid : null,
              ]}
            >
              Create
            </Text>
          </Card>
        </TouchableOpacity>
      </View>
    );
  }
}

CreateExerciseScreen.propTypes = {
  navigation: PropTypes.shape({
    navigate: PropTypes.func.isRequired,
  }).isRequired,
  addExerciseDispatch: PropTypes.func.isRequired,
};

const mapDispatchToProps = (dispatch) => ({
  addExerciseDispatch: (exercise) => {
    dispatch(addExerciseAction(exercise));
  },
});

export default connect(null, mapDispatchToProps)(CreateExerciseScreen);
