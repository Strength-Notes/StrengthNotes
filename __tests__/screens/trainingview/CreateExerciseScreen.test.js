import React from 'react';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import CreateExerciseScreen from '../../../screens/trainingview/CreateExerciseScreen';

const mockStore = configureMockStore();
const store = mockStore({});

describe('<CreateExerciseScreen />', () => {
  it('renders correctly', () => {
    const tree = renderer.create(
      <Provider store={store}>
        <CreateExerciseScreen
          navigation={{ navigate: jest.fn() }}
        />
      </Provider>,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
