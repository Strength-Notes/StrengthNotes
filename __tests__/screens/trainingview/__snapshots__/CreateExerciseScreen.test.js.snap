// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CreateExerciseScreen /> renders correctly 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "backgroundColor": "white",
        "borderColor": "#e1e8ee",
        "borderWidth": 1,
        "margin": 15,
        "marginBottom": 0,
        "padding": 15,
        "shadowColor": "rgba(0,0,0, .2)",
        "shadowOffset": Object {
          "height": 0,
          "width": 0,
        },
        "shadowOpacity": 1,
        "shadowRadius": 1,
      }
    }
  >
    <View
      style={
        Object {
          "backgroundColor": "transparent",
        }
      }
    >
      <View>
        <Text
          style={
            Object {
              "color": "#43484d",
              "fontSize": 17.5,
              "fontWeight": "bold",
              "marginBottom": 15,
              "textAlign": "center",
            }
          }
          testID="cardTitle"
          theme={
            Object {
              "colors": Object {
                "disabled": "hsl(208, 8%, 90%)",
                "divider": "#bcbbc1",
                "error": "#ff190c",
                "grey0": "#393e42",
                "grey1": "#43484d",
                "grey2": "#5e6977",
                "grey3": "#86939e",
                "grey4": "#bdc6cf",
                "grey5": "#e1e8ee",
                "greyOutline": "#bbb",
                "platform": Object {
                  "android": Object {
                    "error": "#f44336",
                    "primary": "#2196f3",
                    "secondary": "#9C27B0",
                    "success": "#4caf50",
                    "warning": "#ffeb3b",
                  },
                  "ios": Object {
                    "error": "#ff3b30",
                    "primary": "#007aff",
                    "secondary": "#5856d6",
                    "success": "#4cd964",
                    "warning": "#ffcc00",
                  },
                },
                "primary": "#2089dc",
                "searchBg": "#303337",
                "secondary": "#8F0CE8",
                "success": "#52c41a",
                "warning": "#faad14",
              },
            }
          }
        >
          Name
        </Text>
        <View
          style={
            Object {
              "backgroundColor": "#bcbbc1",
              "height": 0.5,
              "marginBottom": 15,
            }
          }
        />
      </View>
      <TextInput
        allowFontScaling={true}
        onChangeText={[Function]}
        placeholder="Name"
        rejectResponderTermination={true}
        underlineColorAndroid="transparent"
      />
    </View>
  </View>
  <View
    style={
      Object {
        "backgroundColor": "white",
        "borderColor": "#e1e8ee",
        "borderWidth": 1,
        "margin": 15,
        "marginBottom": 0,
        "padding": 15,
        "shadowColor": "rgba(0,0,0, .2)",
        "shadowOffset": Object {
          "height": 0,
          "width": 0,
        },
        "shadowOpacity": 1,
        "shadowRadius": 1,
      }
    }
  >
    <View
      style={
        Object {
          "backgroundColor": "transparent",
        }
      }
    >
      <View>
        <Text
          style={
            Object {
              "color": "#43484d",
              "fontSize": 17.5,
              "fontWeight": "bold",
              "marginBottom": 15,
              "textAlign": "center",
            }
          }
          testID="cardTitle"
          theme={
            Object {
              "colors": Object {
                "disabled": "hsl(208, 8%, 90%)",
                "divider": "#bcbbc1",
                "error": "#ff190c",
                "grey0": "#393e42",
                "grey1": "#43484d",
                "grey2": "#5e6977",
                "grey3": "#86939e",
                "grey4": "#bdc6cf",
                "grey5": "#e1e8ee",
                "greyOutline": "#bbb",
                "platform": Object {
                  "android": Object {
                    "error": "#f44336",
                    "primary": "#2196f3",
                    "secondary": "#9C27B0",
                    "success": "#4caf50",
                    "warning": "#ffeb3b",
                  },
                  "ios": Object {
                    "error": "#ff3b30",
                    "primary": "#007aff",
                    "secondary": "#5856d6",
                    "success": "#4cd964",
                    "warning": "#ffcc00",
                  },
                },
                "primary": "#2089dc",
                "searchBg": "#303337",
                "secondary": "#8F0CE8",
                "success": "#52c41a",
                "warning": "#faad14",
              },
            }
          }
        >
          Primary
        </Text>
        <View
          style={
            Object {
              "backgroundColor": "#bcbbc1",
              "height": 0.5,
              "marginBottom": 15,
            }
          }
        />
      </View>
      <View>
        <RCTPicker
          items={
            Array [
              Object {
                "label": "Weight",
                "textColor": undefined,
                "value": "Weight",
              },
              Object {
                "label": "Reps",
                "textColor": undefined,
                "value": "Reps",
              },
              Object {
                "label": "Time",
                "textColor": undefined,
                "value": "Time",
              },
              Object {
                "label": "Distance",
                "textColor": undefined,
                "value": "Distance",
              },
              Object {
                "label": "None",
                "textColor": undefined,
                "value": "None",
              },
            ]
          }
          onChange={[Function]}
          selectedIndex={0}
          style={
            Array [
              Object {
                "height": 216,
              },
              undefined,
            ]
          }
        />
      </View>
    </View>
  </View>
  <View
    style={
      Object {
        "backgroundColor": "white",
        "borderColor": "#e1e8ee",
        "borderWidth": 1,
        "margin": 15,
        "marginBottom": 0,
        "padding": 15,
        "shadowColor": "rgba(0,0,0, .2)",
        "shadowOffset": Object {
          "height": 0,
          "width": 0,
        },
        "shadowOpacity": 1,
        "shadowRadius": 1,
      }
    }
  >
    <View
      style={
        Object {
          "backgroundColor": "transparent",
        }
      }
    >
      <View>
        <Text
          style={
            Object {
              "color": "#43484d",
              "fontSize": 17.5,
              "fontWeight": "bold",
              "marginBottom": 15,
              "textAlign": "center",
            }
          }
          testID="cardTitle"
          theme={
            Object {
              "colors": Object {
                "disabled": "hsl(208, 8%, 90%)",
                "divider": "#bcbbc1",
                "error": "#ff190c",
                "grey0": "#393e42",
                "grey1": "#43484d",
                "grey2": "#5e6977",
                "grey3": "#86939e",
                "grey4": "#bdc6cf",
                "grey5": "#e1e8ee",
                "greyOutline": "#bbb",
                "platform": Object {
                  "android": Object {
                    "error": "#f44336",
                    "primary": "#2196f3",
                    "secondary": "#9C27B0",
                    "success": "#4caf50",
                    "warning": "#ffeb3b",
                  },
                  "ios": Object {
                    "error": "#ff3b30",
                    "primary": "#007aff",
                    "secondary": "#5856d6",
                    "success": "#4cd964",
                    "warning": "#ffcc00",
                  },
                },
                "primary": "#2089dc",
                "searchBg": "#303337",
                "secondary": "#8F0CE8",
                "success": "#52c41a",
                "warning": "#faad14",
              },
            }
          }
        >
          Secondary
        </Text>
        <View
          style={
            Object {
              "backgroundColor": "#bcbbc1",
              "height": 0.5,
              "marginBottom": 15,
            }
          }
        />
      </View>
      <View>
        <RCTPicker
          items={
            Array [
              Object {
                "label": "Weight",
                "textColor": undefined,
                "value": "Weight",
              },
              Object {
                "label": "Reps",
                "textColor": undefined,
                "value": "Reps",
              },
              Object {
                "label": "Time",
                "textColor": undefined,
                "value": "Time",
              },
              Object {
                "label": "Distance",
                "textColor": undefined,
                "value": "Distance",
              },
              Object {
                "label": "None",
                "textColor": undefined,
                "value": "None",
              },
            ]
          }
          onChange={[Function]}
          selectedIndex={1}
          style={
            Array [
              Object {
                "height": 216,
              },
              undefined,
            ]
          }
        />
      </View>
    </View>
  </View>
  <View
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <View
      style={
        Object {
          "backgroundColor": "lightgray",
          "borderColor": "#e1e8ee",
          "borderWidth": 1,
          "margin": 15,
          "marginBottom": 0,
          "padding": 15,
          "shadowColor": "rgba(0,0,0, .2)",
          "shadowOffset": Object {
            "height": 0,
            "width": 0,
          },
          "shadowOpacity": 1,
          "shadowRadius": 1,
        }
      }
    >
      <View
        style={
          Object {
            "backgroundColor": "transparent",
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "textAlign": "center",
              },
              Object {
                "color": "gray",
              },
            ]
          }
        >
          Create
        </Text>
      </View>
    </View>
  </View>
</View>
`;
